I want to find out how many episodes of Spongebob Squarepants the charcter Patrick Star appears in.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('/content/spongebob_episodes.csv')

In [3]:
df

Airdate                                   Animation  \
0           1997/1998                                  Alan Smart   
1         May 1, 1999  ['Fred Miller', 'Tom Yasumi (uncredited)']   
2         May 1, 1999                                  Tom Yasumi   
3       July 17, 1999                                  Tom Yasumi   
4       July 17, 1999                            Edgar Larrazábal   
..                ...                                         ...   
578     July 22, 2024                                         NaN   
579     July 23, 2024                             Andrew Overtoom   
580     July 24, 2024                             Andrew Overtoom   
581     July 25, 2024                              Michelle Bryan   
582  October 29, 2024                                         NaN   

    Animation Supervisor                                  Copyright year  \
0                    NaN  ['1997 (1997 version)', '1999 (1999 version)']   
1                    NaN                                            1999   
2                    NaN                                            1999   
3                    NaN                                            1999   
4                    NaN                                            1999   
..                   ...                                             ...   
578                  NaN                                            2024   
579                  NaN                                            2024   
580                  NaN                                            2024   
581                  NaN                                            2024   
582                  NaN                                            2024   

                      Creative Episode №                            Guest(s)  \
0    Derek Drymon (uncredited)        1a                                 NaN   
1    Derek Drymon (uncredited)        1b                                 NaN   
2                 Derek Drymon        1c                                 NaN   
3                 Derek Drymon        2a                                 NaN   
4                 Derek Drymon        2b     Jay Lender as Chef (uncredited)   
..                         ...       ...                                 ...   
578                        NaN      304a                                 NaN   
579                        NaN      304b                                 NaN   
580                        NaN      305a  Richard Ayoade as Sammy Suckerfish   
581                        NaN      305b                                 NaN   
582                        NaN       306                                 NaN   

    Line Producer Main                                 Next  ...  \
0             NaN  NaN                        "Reef Blower"  ...   
1             NaN  NaN                "Tea at the Treedome"  ...   
2             NaN  NaN                        "Bubblestand"  ...   
3             NaN  NaN                       "Ripped Pants"  ...   
4             NaN  NaN                       "Jellyfishing"  ...   
..            ...  ...                                  ...  ...   
578   Joe Crowley  NaN                "Sheldon SquarePants"  ...   
579   Joe Crowley  NaN                   "Sammy Suckerfish"  ...   
580   Joe Crowley  NaN                     "Big League Bob"  ...   
581   Joe Crowley  NaN  "Snow Yellow and the Seven Jellies"  ...   
582   Joe Crowley  NaN                             "UpWard"  ...   

                    Storyboard     Storyboard Artist(s)      Supervising  \
0           Stephen Hillenburg             Derek Drymon              NaN   
1    Paul Tibbitt (uncredited)  Jay Lender (uncredited)              NaN   
2                 Paul Tibbitt              Mark O'Hare              NaN   
3               Ennio Torresan               Erik Wiese              NaN   
4                 Paul Tibbitt              Mark O'Hare              NaN   
..                         ...        

In [4]:
df.head()

Airdate                                   Animation  \
0      1997/1998                                  Alan Smart   
1    May 1, 1999  ['Fred Miller', 'Tom Yasumi (uncredited)']   
2    May 1, 1999                                  Tom Yasumi   
3  July 17, 1999                                  Tom Yasumi   
4  July 17, 1999                            Edgar Larrazábal   

  Animation Supervisor                                  Copyright year  \
0                  NaN  ['1997 (1997 version)', '1999 (1999 version)']   
1                  NaN                                            1999   
2                  NaN                                            1999   
3                  NaN                                            1999   
4                  NaN                                            1999   

                    Creative Episode №                         Guest(s)  \
0  Derek Drymon (uncredited)        1a                              NaN   
1  Derek Drymon (uncredited)        1b                              NaN   
2               Derek Drymon        1c                              NaN   
3               Derek Drymon        2a                              NaN   
4               Derek Drymon        2b  Jay Lender as Chef (uncredited)   

  Line Producer Main                   Next  ...                 Storyboard  \
0           NaN  NaN          "Reef Blower"  ...         Stephen Hillenburg   
1           NaN  NaN  "Tea at the Treedome"  ...  Paul Tibbitt (uncredited)   
2           NaN  NaN          "Bubblestand"  ...               Paul Tibbitt   
3           NaN  NaN         "Ripped Pants"  ...             Ennio Torresan   
4           NaN  NaN         "Jellyfishing"  ...               Paul Tibbitt   

      Storyboard Artist(s) Supervising  Supervising Producer(s) Technical  \
0             Derek Drymon         NaN                      NaN       NaN   
1  Jay Lender (uncredited)         NaN                      NaN       NaN   
2              Mark O'Hare         NaN                      NaN       NaN   
3               Erik Wiese         NaN                      NaN       NaN   
4              Mark O'Hare         NaN                      NaN       NaN   

  U.S. premiere time (ET) U.S. viewers (millions)  \
0                 9:30 pm                    2.96   
1                 9:38 pm                    2.96   
2                 9:45 pm                    2.96   
3                10:00 am                    2.51   
4                10:15 am                    2.51   

                                           Writer(s)  \
0  ['Stephen Hillenburg', 'Derek Drymon', 'Tim Hi...   
1  ['Stephen Hillenburg', 'Derek Drymon', 'Tim Hi...   
2   ['Peter Burns', 'Doug Lawrence', 'Paul Tibbitt']   
3  ['Ennio Torresan', 'Erik Wiese', 'Stephen Hill...   
4                    ['Paul Tibbitt', 'Peter Burns']   

                                          characters                title  
0  French Narrator, Sea urchins, SpongeBob Square...          Help Wanted  
1  SpongeBob SquarePants, Squidward Tentacles, Re...          Reef Blower  
2  SpongeBob SquarePants, Jellyfish, Lavender jel...  Tea at the Treedome  
3  SpongeBob SquarePants, Scallop, Squidward Tent...          Bubblestand  
4  French Narrator, SpongeBob SquarePants, Sponge...         Ripped Pants  

[5 rows x 25 columns]

In [5]:
if 'characters' not in df.columns or 'title' not in df.columns:
    print("Error: 'characters' or 'title' column not found. Please check the dataset.")
    exit()

In [6]:
def process_character_string(char_str):
    try:
        if isinstance(char_str, list): # Already a list, return as is
          return char_str
        if pd.isna(char_str) or not isinstance(char_str, str):  # handle NaN or other non-string values
            return []
        char_str = char_str.replace('[', '').replace(']', '') #Remove brackets
        char_list = [s.strip("' ") for s in char_str.split(',')] #Remove quotes and spaces
        return [s for s in char_list if s] #filter out empty strings
    except Exception as e:
      print(f"Error processing character string: {char_str}, Error: {e}")
      return [] # Return an empty list to avoid breaking the rest of the code.


df['characters'] = df['characters'].apply(process_character_string)

In [7]:
patrick_episodes = df[df['characters'].apply(lambda x: 'Patrick Star' in x)]
num_patrick_episodes = len(patrick_episodes)

print(f"Patrick Star appears in {num_patrick_episodes} episodes.")

Patrick Star appears in 434 episodes.


In [9]:
print("\nEpisodes with Patrick Star:")
print(patrick_episodes['title'])


Episodes with Patrick Star:
0                Help Wanted
2        Tea at the Treedome
3                Bubblestand
5               Jellyfishing
6                  Plankton!
               ...          
576       Hysterical History
577           Kreepaway Kamp
578    The Dirty Bubble Bass
579      Sheldon SquarePants
581           Big League Bob
Name: title, Length: 434, dtype: object


It looks like Patrick Star appears in 434 episodes out of the 583 total. Thats quite a bit!